compiler/types.nim

```diff
@@ -1471,6 +1471,35 @@ proc skipHiddenSubConv*(n: PNode; g: ModuleGraph; idgen: IdGenerator): PNode =
   else:
     result = n
 
+proc callConvMismatch*(formal, actual: PType, indented = false): string =
+  assert formal.kind == tyProc and actual.kind == tyProc
+  let ambigNimProc = formal.callConv in {ccNimCall, ccClosure} and tfExplicitCallConv notin formal.flags
+  if (ambigNimProc and actual.callConv notin {ccNimCall, ccClosure}) or
+      (not ambigNimProc and formal.callConv != actual.callConv):
+    let
+      got = $(actual.callConv)
+      expected = $(formal.callConv)
+    result = "\n  Calling convention: mismatch got '{.$1.}', but expected '{.$2.}'." % [got, expected]
+
+
+proc pragmaMismatch*(formal, actual: PType): string =
+  assert formal.kind == tyProc and actual.kind == tyProc
+  var
+    got = ""
+    expected = ""
+  if tfNoSideEffect in formal.flags and tfNoSideEffect notin actual.flags:
```

Comment:
option question: can we do some refactor to avoid duplicating logic in sigmatch.procTypeRel so that it stays in sync?

Reply:
Well it does make sense but sigmatch imports types so `TTypeRelation` will have to get moved to types and the procedure exist in there, though I did start on this.

Reply:
maybe try on a separate PR to de-risk this approach?

Reply:
Well i have a branch now with the less redundant method for this if to take a look at, it's mergable here if wanted.
https://github.com/beef331/Nim/commit/38f0341f52a9c3139e4c706c1285944476e94bfe

Reply:
excellent, exactly what i had in mind, and even better since you also return `set[ProcConvMismatch]` instead of just the 1st reason `ProcConvMismatch`; I'm assuming this is ok performance wise because getProcConvMismatch isn't a hotspot at least in the case where things don't match, so it's ok to do a tiny bit more work there.